packages/block-editor/src/hooks/use-bindings-attributes.js

```diff
@@ -22,89 +24,206 @@ import { unlock } from '../lock-unlock';
+	args,
+	attrName,
+	attrValue,
+	blockProps,
+	source,
+} ) => {
+	const { placeholder, value: propValue } = source.useSource(
+		blockProps,
+		args
+	);
+
+	const blockName = blockProps.name;
+
+	const setAttributes = blockProps.setAttributes;
+
+	const updateBoundAttibute = useCallback(
+		( newAttrValue ) => {
+			/*
+			 * If the attribute is a RichTextData instance,
+			 * (core/paragraph, core/heading, etc.)
+			 * convert it to HTML string and compare with the new value.
+			 * If they are the same, don't update the attribute.
+			 *
+			 * To do: it looks like a workaround.
+			 * Consider improving the attribute and metadata fields types.
+			 */
+			if (
+				attrValue instanceof RichTextData &&
+				attrValue.toHTMLString() === newAttrValue
+			) {
+				return;
 			}
 
-			return (
-				<BlockEdit
-					key="edit"
-					{ ...props }
-					attributes={ updatedAttributes }
-				/>
-			);
+			setAttributes( {
+				[ attrName ]: newAttrValue,
+			} );
 		},
-		'useBoundAttributes'
+		[ attrName, attrValue, setAttributes ]
 	);
 
+	useEffect( () => {
+		if ( typeof propValue !== 'undefined' ) {
+			updateBoundAttibute( propValue );
+		} else if ( placeholder ) {
+			/*
+			 * Placeholder fallback.
+			 * If the attribute is `src` or `href`,
+			 * a placeholder can't be used because it is not a valid url.
+			 * Adding this workaround until
+			 * attributes and metadata fields types are improved and include `url`.
+			 */
+			const htmlAttribute =
+				getBlockType( blockName ).attributes[ attrName ].attribute;
+
+			if ( htmlAttribute === 'src' || htmlAttribute === 'href' ) {
+				updateBoundAttibute( null );
+				return;
+			}
+
+			updateBoundAttibute( placeholder );
+		}
+	}, [
+		updateBoundAttibute,
+		propValue,
+		attrValue,
+		placeholder,
+		blockName,
+		attrName,
+	] );
+
+	return null;
+};
+
+function BlockBindingBridge( { bindings, props } ) {
+	const { getBlockBindingsSource } = useSelect( () => {
+		return {
+			getBlockBindingsSource: unlock( select( blockEditorStore ) )
+				.getBlockBindingsSource,
+		};
+	}, [] );
+
+	if ( ! bindings ) {
+		return null;
+	}
+
+	const { name, attributes } = props;
+
+	// Collect all the binding connectors.
+	const BindingConnectorInstances = [];
+
+	Object.entries( bindings ).forEach( ( [ attrName, boundAttribute ], i ) => {
+		// Check if the block attribute can be bound.
+		if ( ! hasPossibleBlockBinding( name, attrName ) ) {
+			return;
+		}
+
+		// Bail early if the block doesn't have a valid source handler.
+		const source = getBlockBindingsSource( boundAttribute.source );
+		if ( ! source?.useSource ) {
+			return;
+		}
+
+		BindingConnectorInstances.push(
+			<BlockBindingConnector
+				key={ `${ boundAttribute.source }-${ name }-${ attrName }-${ i }` }
+				attrName={ attrName }
+				attrValue={ attributes[ attrName ] }
+				source={ source }
+				blockProps={ props }
+				args={ boundAttribute.args }
+			/>
+		);
+	} );
+
+	return <>{ BindingConnectorInstances }</>;
```

Comment:
This could also be done in a `.map` that outputs the react elements rather than pushing them onto an array.

Creating all these `BlockConnector`s  per attribute seems like a lot and has me worried about performance. I guess its possible for each attribute to be bound to a different source, so something like this is needed.

Reply:
Yes, we used the `< BlockBindingConnector />` component instance. The external source of an attribute is arbitrary and potentially unrelated to the other bound ones.